### PR TITLE
Spy support

### LINKED ISSUE #33: spy support
Hey guys, 
Any plan on supporting spy in the same style as java mockito does? the library is fantastic by the way
Kind Regards
Gabriel

### PR DESCRIPTION
**Pull request by dreef3**
Hi,

This is an attempt to reuse the existing `Mocker` class to provide a functionality for spying on the existing object.

Closes #33

### CODE REVIEW

**Comment by NagRock** [comment]
@dreef3 thanks for another pull request! I will check it and give feedback.

**Comment by dreef3** [comment]
@NagRock this one definitely needs more tests, will add them if the approach to implementation looks good.

File: test/spy.spec.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        get baz() {
+            return 3;
+        }
+    }
+
+    describe("calling a mocked method", () => {
+        it("delegates the call to a mock", () => {
+            let foo = new Real();
```
Comment (NagRock) [inline test/spy.spec.ts]: please use given/when/then convention

File: README.md [hunk 2]
```diff
@@ -306,6 +307,18 @@ const foo: SampleGeneric<SampleInterface> = instance(mockedFoo);
 
 ```
 
+### Spying on real objects
+
+You can partially mock an existing instance:
+
+``` typescript
+const foo: Foo = new Foo();
+const mockedFoo = spy(foo);
```
Comment (NagRock) [inline README.md]: I think that `spiedFoo` would be better name. What do you think?

File: test/spy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            expect(capture(mocked.bar).last()).toBeDefined();
+        });
+
+        it("captures the call arguments", () => {
+            let foo = new Real();
+            let mocked = spy(foo);
+
+            foo.foo(42);
+
+            expect(capture(mocked.foo).last()).toEqual([42]);
```
Comment (NagRock) [inline test/spy.spec.ts]: Why we need to use array `[42]` here instead of number `42`?

File: test/spy.spec.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        });
+    });
+
+    describe("resetting", () => {
+        it("restores the call to a real method", () => {
```
Comment (NagRock) [inline test/spy.spec.ts]: Nice test! 👍

**Comment by NagRock** [comment]
Nice that `calledBefore` and `calledAfter` functionality still works!

I need some more time to check API, because when using mock we got to use `instance` while using `spy` we're not. Please give me one more day to check it :) And one more time thanks for contributing!

File: test/spy.spec.ts (hunk 3 above)
Comment (dreef3) [reply test/spy.spec.ts]: Because `ArgCaptor.last()` returns an array of arguments

**Comment by NagRock** [comment]
@dreef3 hi again, it looks great! Two things. We need some more tests (for example call order testing `calledBefore`, `calledAfter` and call count `once()`, `twice()` etc.)

And second thing:

```
            let foo = new Real();
            let mocked = spy(foo);

            when(mocked.foo(1)).thenReturn(3);

            expect(foo.foo(1)).toBe(3);
            expect(foo.foo(2)).toBe(2); // here should be called real implementation but we get null
```

If you need some more information or help with something then ping me!

File: src/Mock.ts [hunk 5]
```diff
@@ -59,36 +58,46 @@ export class Mocker {
...
 
-            prototype = prototype.__proto__;
-            if (prototype && prototype !== Object.prototype) {
-                this.createMethodStubsFromPrototypeOwnPropertyNames(prototype);
+                if (!recurse) {
+                    return;
+                }
+
+                prototype = prototype.__proto__;
+                this.createMethodStubsFromOwnPropertyNames(prototype);
```
Comment (viman) [inline src/Mock.ts]: See the comment above

File: src/Mock.ts [hunk 6]
```diff
@@ -155,27 +164,34 @@ export class Mocker {
...
             }
 
+            if (!recurse) {
+                return;
+            }
+
             prototype = prototype.__proto__;
             if (prototype && prototype !== Object.prototype) {
-                this.createInstanceActionListenersFromPrototypeOwnPropertyDescriptors(prototype);
+                this.createInstanceActionListenersFromOwnPropertyDescriptors(prototype);
```
Comment (viman) [inline src/Mock.ts]: If you are changing recursion into while loop please be consistent and implement it here also.

File: src/Mock.ts [hunk 7]
```diff
@@ -185,16 +201,21 @@ export class Mocker {
...
+            }
+
+            if (!recurse) {
+                return;
             }
 
             prototype = prototype.__proto__;
             if (prototype && prototype !== Object.prototype) {
-                this.createInstanceActionListenersFromPrototypeOwnPropertyNames(prototype);
+                this.createInstanceActionListenersFromOwnPropertyNames(prototype);
```
Comment (viman) [inline src/Mock.ts]: See the comment above.

File: src/spy/RealMethod.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export class RealMethod {
+    constructor(public descriptor: PropertyDescriptor,
+                public instance: boolean) {}
+}
```
Comment (viman) [inline src/spy/RealMethod.ts]: Please add a new line on the end of files.

**Comment by viman** [comment]
Great job @dreef3 with the spies! 👍 Please see my comments.

File: src/spy/RealMethod.ts (hunk 8 above)
Comment (dreef3) [reply src/spy/RealMethod.ts]: done

File: src/Mock.ts [hunk 9]
```diff
@@ -59,36 +58,46 @@ export class Mocker {
...
 
-            prototype = prototype.__proto__;
-            if (prototype && prototype !== Object.prototype) {
-                this.createMethodStubsFromPrototypeOwnPropertyDescriptors(prototype);
+                if (!recurse) {
+                    return;
+                }
+
+                prototype = prototype.__proto__;
+                this.createMethodStubsFromOwnPropertyDescriptors(prototype);
```
[the remark this answers is not in the document]
Comment (dreef3) [reply src/Mock.ts]: Reverted to recursion, merged two methods into one

**Comment by dreef3** [comment]
@NagRock,

1. I've added some more tests for verification, not sure it's necessary to test all verification methods because it would duplicate existing `verification.spec.ts`.
2. Fixed this case

File: src/Mock.ts (hunk 6 above)
Comment (dreef3) [reply src/Mock.ts]: Reverted the while loop, only needed to check for Object.prototype at the beginning to catch the case with plain object without inheritance hierarchy.

**Comment by NagRock** [comment]
@dreef3 Thanks for changes! You right that it would duplicate existing specs so don't do this. I will try to parameterise it later.

**Comment by NagRock** [comment]
@dreef3 merged! Once again thanks for this feature!
